File: bot-frontend/src/features/rpc-groups/hooks/rpcGroups.ts
import { useMutation, useQuery, useQueryClient } from '@tanstack/react-query';
import {
  deleteRPCGroup,
  getRPCDashboard,
  getRPCGroup,
  postRPCGroup,
  updateRPCGroup,
} from '../api/rpcGroups';
import type { RPCGroupPost, RPCGroupPut } from '../types/rpcGroup';
import { toast } from 'sonner';

export function useRpcGroupDashboard() {
  return useQuery({
    queryKey: ['rpcGroupDashboard'],
    queryFn: getRPCDashboard,
  });
}

export function usePostRPCGroup() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (rpcGroup: RPCGroupPost) => postRPCGroup(rpcGroup),
    onSuccess: () => {
      client.invalidateQueries({ queryKey: ['rpcGroupDashboard'] });
    },
  });
}

export function useRpcGroup({ id }: { id: number }) {
  return useQuery({
    queryKey: ['rpcGroup', id],
    queryFn: () => getRPCGroup(id),
  });
}

export function useUpdateRPCGroup() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (rpcGroup: RPCGroupPut) => updateRPCGroup(rpcGroup),
    onSuccess: (_, variables) => {
      client.invalidateQueries({ queryKey: ['rpcGroupDashboard'] });
      client.invalidateQueries({ queryKey: ['rpcGroup', variables.id] });
    },
  });
}

export function useDeleteRPCGroup() {
  const client = useQueryClient();
  return useMutation({
    mutationFn: (id: number) => deleteRPCGroup(id),
    onSuccess: (_, id) => {
      client.invalidateQueries({ queryKey: ['rpcGroupDashboard'] });
      client.removeQueries({ queryKey: ['rpcGroup', id] });
    },
    onError: (e) => toast.error('Failure to delete RPC Group', { description: e.message }),
  });
}

File: backend/internal/core/rpc_groups/rpc_groups.go
package rpcgroups

import "github.com/gagliardetto/solana-go/rpc"

type RPCGroup struct {
	Id           int64
	Name         string
	Group        Group
	CreationTime string
}

type LoadedRPCGroup struct {
	RPCGroup   RPCGroup
	GroupIndex int
	References int
}

type RPCGroupPost struct {
	Name         string
	Group        Group
	CreationTime string
}

type RPCGroupPut struct {
	Name  string
	Group Group
}

type Group []GroupItem

type RPCNode struct {
	Http *rpc.Client
	WS   string
}

type GroupItem struct {
	Http string `json:"http"`
	WS   string `json:"ws"`
}

type RPCGroupDashboard []RPCGroupDashboardRow

type RPCGroupDashboardRow struct {
	Id           int64  `json:"id"`
	Name         string `json:"name"`
	Number       int64  `json:"number"`
	CreationTime string `json:"creation_time"`
}
